File: 2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points/2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public int[] nodesBetweenCriticalPoints(ListNode head) {
        List<Integer> tmp=new ArrayList<>();
        int ans[]=new int[]{-1,-1};
        if(head==null || head.next==null || head.next.next==null)
            return ans;
        ListNode curr=head.next,prev=head,next=head.next.next;
        int max=Integer.MIN_VALUE,min=Integer.MAX_VALUE;
        int idx=2;
        while(next!=null){
            if((curr.val<prev.val && curr.val<next.val)|| (curr.val>prev.val && curr.val>next.val))
                tmp.add(idx);
            idx++;
            prev=curr;
            curr=next;
            next=next.next;
        }
        // for(int i=0;i<tmp.size();i++){
        //     for(int j=0;j<tmp.size();j++){
        //         if(i!=j){
        //             max=Math.max(max,Math.abs(tmp.get(i)-tmp.get(j)));
        //             min=Math.min(min,Math.abs(tmp.get(i)-tmp.get(j)));
        //         }
        //     }
        // }
        if(tmp.size()>=2)
            max=tmp.get(tmp.size()-1)-tmp.get(0);
        for(int i=0;i<tmp.size()-1;i++){
            min=Math.min(min,tmp.get(i+1)-tmp.get(i));
        }
        if(min!=Integer.MAX_VALUE && max!=Integer.MIN_VALUE){
            ans[0]=min;
            ans[1]=max;
        }
        
        //System.out.println(tmp);
        return ans;
    }
}

File: 2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points/2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def nodesBetweenCriticalPoints(self, head):
        """
        :type head: Optional[ListNode]
        :rtype: List[int]
        """
        l,ans,idx=[],[-1,-1],2
        if not head or not head.next or not head.next.next:
            return ans

        curr,prev,next_=head.next,head,head.next.next

        while next_:
            if (curr.val>prev.val and curr.val>next_.val) or (curr.val<prev.val and curr.val<next_.val):
                l.append(idx)
            idx=idx+1
            prev=curr
            curr=next_
            next_=next_.next
        
        max,min=-1,pow(10,6)
        # print(l)
        if len(l)>=2:
            max=l[-1]-l[0]
        for i in range(len(l)-1):
            if min>(l[i+1]-l[i]):
                min=l[i+1]-l[i]
        if max!=-1 and min!=pow(10,6):
            ans[0]=min
            ans[1]=max

        return ans
